consensus/src/chained_bft/consensus_types/mod.rs

```diff
@@ -1,8 +1,40 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
+//! The `mod.rs` of this module sets the types containing cryptographic
+//! materials used in the whole consensus crate. It is intended as a
+//! one-stop shop for changing the signing scheme of the consensus. If
+//! consensus uses a type that contains cryptographic material, its
+//! polymorphic parameter should be instantiated here and re-exported.
+//!
+//! Imports specific to any signing scheme (e.g. ed25519::*) should not be
+//! present anywhere else in the consensus crate. Use of raw cryptographic
+//! types that do not go through the instantiated polymorphic structures
+//! below is banned.
+//!
+//! For more on the policy of use of cryptographic material in consensus,
+//! including insight on genericity, please have a look at
+//! `crate::chained_bft::structs`.
+
 pub(crate) mod block;
 pub(crate) mod proposal_msg;
 pub(crate) mod quorum_cert;
 pub(crate) mod sync_info;
 pub(crate) mod timeout_msg;
+
+use crate::chained_bft::structs::SignatureWrapper;
+use crypto::ed25519::*;
+use types::{
+    ledger_info::LedgerInfoWithSignatures as RawLedgerInfoWithSignatures,
+    validator_signer::ValidatorSigner as RawValidatorSigner,
+    validator_verifier::ValidatorVerifier as RawValidatorVerifier,
+};
+
+// used in chained_bft::consensus_types::block_test
+#[cfg(test)]
+pub(crate) type SecretKey = Ed25519PrivateKey;
+
+pub type Signature = SignatureWrapper<Ed25519Signature>;
```

Comment:
For lines 37-40, other parts of the system are using the original types with generics directly.  Does it make more sense to push the type aliases into the crypto crate and then convert the entire codebase to using the type aliases instead?  I know that consensus will use a different signature scheme than transactions for example, but we could type alias ConsensusSignature and TransactionSignature for example.

I'm a little worried about having a different pattern in consensus vs the rest of the code base.

Reply:
> other parts of the system 
Non-consensus ones. 

> Does it make more sense to push the type aliases into the crypto crate and then convert the entire codebase to using the type aliases instead? 

As discussed between us, no, for three reasons:
- those low-level binding structs are a scheme choice boundary, as you correctly inferred from the "ConsensusSignature vs TransactionSignature" remark. The crypto crate is not a good place to maintain those boundaries in a reasonable manner (in particular, we'd need to manage the `NetworkPublicKey` alias today ... and more tomorrow).
- the conventions that come with those boundaries (only use structs that depend on your local aliases in the crate) would become more brittle and harder to maintain, as the crypto crate / pod does not have context on each of the set of structs in each area (consensus, network, language, vm, storage ...)
- you can't define all of this in the crypto crate, as other cryptographic structures those wrappers need to interact with (e.g. `Validator(Signer|Verifier), LedgerInfoWithSigantures`) are defined *in function of* the crypto crate (aka depend on it).

> I'm a little worried about having a different pattern in consensus vs the rest of the code base.

I can rebase #522 in a few h if you change your mind :)

Reply:
I agree that crypto might not be the right place as it shouldn't have any client specific knowledge.  However, types might be a good place instead - all these structs exist there as well.  For example, with LedgerInfoWithSignatures it would be nice for storage/execution to use it as well.  Same with ValidatorVerifier, ValidatorSigner, and  Signature.

Reply:
This addresses point 3, but 1 & 2 are still something we'll have to discuss.